src/handlers/proposalVersionsHandlers.ts

```diff
@@ -0,0 +1,224 @@
+import { getLogger } from '../logger';
+import { db } from '../database';
+import {
+  isApplicationForm,
+  isApplicationFormField,
+  isProposal,
+  isProposalVersionWrite,
+  isProposalVersion,
+  isProposalFieldValue,
+  isTinyPgErrorWithQueryContext,
+} from '../types';
+import {
+  DatabaseError,
+  InternalValidationError,
+  InputValidationError,
+  InputConflictError,
+} from '../errors';
+import type {
+  Request,
+  Response,
+  NextFunction,
+} from 'express';
+import type {
+  ApplicationForm,
+  ApplicationFormField,
+  Proposal,
+  ProposalVersion,
+  ProposalVersionWrite,
+  ProposalFieldValue,
+  ProposalFieldValueWrite,
+} from '../types';
+
+const logger = getLogger(__filename);
+
+const assertApplicationFormExistsForProposal = async (
```

Comment:
On the one hand, this is a good validation and it looks like it keeps things correctly synchronized. On the other hand, it seems like the relations are wonky if the validation is needed. I probably had the same question about this in the past that I cannot find in GH comments, sorry, so I'll ask again: what would go missing if a proposal was in response to an application form rather than in response to an opportunity? The opportunity can be deduced from the application form and the possibility of mixing things up vanishes. A similar thing might be possible for application form fields, though (posting a field value and field that does not belong to that form), and I cannot come up with a similar suggestion there.

I have a vague memory of you suggesting that we would relate proposalVersion to applicationForm and that is done here and is probably fine.

Reply:
So, ERD aside, at a conceptual level a proposal (the concept of "hey I want to apply!") is always in response to an opportunity and that a proposal version (the concept of "here is some information!") will always be in responce  to a specific request for information defined within that opportunity (e.g. specific application form).

By linking a *PROPOSAL* to an application form we couple the desire to apply itself to a specific request for information as opposed to the more general opportunity.  This doesn't necessarily result in lost data on surface, but if we think about potential life cycles of the data I think it becomes a bit more clear: if someone wanted to delete an obsolete application form that would break relationships between a proposal and an opportunity if we link through that entity. (We won't be doing deletions at this point, but from a ERD perspective I think that helps clarify?)

A non-deletion example: if an opportunity updates its application forms and wants to collect some additional information / adds a question to an application form, this would appropriately make a proposal *version* obsolete, but the *proposal* is still in fine order (it's just that it doesn't yet have a version that is responsive to the most recent request for information / application form).

With the current model the proposal is linked to an opportunity (which may have any number of application forms / application form versions) and the proposal version is the only thing locked into a specific application form version.

On a related / side note; in one of our feedback sessions the concept of having more than one *ACTIVE* application form associated with an opportunity was raised (e.g. people should fill out form X, Y, and Z which individually might have evolving version!)  This would mean creating another entity under "ApplicationForm" and allowing a single proposal to be responsive to more than one application form (thus making it I think more clear why the opportunity is important, though we don't need to think about this here / this is not the leading reason for the proposed design).

... I am worried that none of the above makes any sense 😭 

---

I don't know if psql allows for foreign key relationships *THROUGH* other tables.  I'll look into that, as it would allow us to avoid this check.

Reply:
That all makes sense, yes, and explains the utility of modeling and relating a proposal to an opportunity separately from application form.

Reply:
I spent a bit of time trying to think through this (and pinging some friends about the broader problem space).  The verdict seems to be there is no great solution.

It would be possible to create a trigger that validates this relationship on insert.  there is also the concept of a `CHECK` but psql documentation is pretty clear that CHECK should not be used to validate anything outside of the columns being directly inserted:

```
PostgreSQL does not support CHECK constraints that reference table data other than the new or updated row being checked. While a CHECK constraint that violates this rule may appear to work in simple tests, it cannot guarantee that the database will not reach a state in which the constraint condition is false (due to subsequent changes of the other row(s) involved). This would cause a database dump and restore to fail. The restore could fail even when the complete database state is consistent with the constraint, due to rows not being loaded in an order that will satisfy the constraint. If possible, use UNIQUE, EXCLUDE, or FOREIGN KEY constraints to express cross-row and cross-table restrictions.

If what you desire is a one-time check against other rows at row insertion, rather than a continuously-maintained consistency guarantee, a custom trigger can be used to implement that. (This approach avoids the dump/restore problem because pg_dump does not reinstall triggers until after restoring data, so that the check will not be 
```

I think that maybe for now the problem might not have a good solution and this runtime check is our best bet (it does mean that, for instance, a proposal version COULD be associated with an invalid application form in the event that the application form's opportunity relationship is changed...)